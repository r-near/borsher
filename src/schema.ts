import { type Schema, deserialize as thirdPartyDeserialize, serialize } from "borsh"

/**
 * Helper type to extract the inner type from a BorshSchema.
 */
export type TypeOf<T extends BorshSchema<unknown, string>> = T extends BorshSchema<infer U, string>
  ? U
  : never

/**
 * BorshSchema is a thin wrapper around the underlying third‐party schema.
 * It carries both compile‐time type information and (via an optional runtime
 * flag) extra metadata that lets our custom deserialize() method convert plain
 * arrays into typed arrays.
 */
export class BorshSchema<T, Tag extends string = never> {
  // The underlying third‐party schema.
  private readonly schema: Schema
  // If set (for vector schemas), indicates that the plain array should be
  // converted into a typed array.
  private readonly _typedArrayType?: string
  // Keep track of type and tag explicitly to help TypeScript with type inference
  private readonly _type: T
  private readonly _tag: Tag

  private constructor(schema: Schema, typedArrayType?: string) {
    this.schema = schema
    this._typedArrayType = typedArrayType
    // These are only used for type inference, values don't matter
    this._type = null as T
    this._tag = "" as Tag
  }

  // Scalar types with exact same definitions as BorshSchema
  static readonly String = new BorshSchema<string>("string")

  // Unsigned integers:
  static readonly u8 = new BorshSchema<number, "u8">("u8")
  static readonly u16 = new BorshSchema<number, "u16">("u16")
  static readonly u32 = new BorshSchema<number, "u32">("u32")
  static readonly u64 = new BorshSchema<bigint, "u64">("u64") // no tag = no special Vec conversion
  static readonly u128 = new BorshSchema<bigint>("u128")

  // Signed integers:
  static readonly i8 = new BorshSchema<number, "i8">("i8")
  static readonly i16 = new BorshSchema<number, "i16">("i16")
  static readonly i32 = new BorshSchema<number, "i32">("i32")
  static readonly i64 = new BorshSchema<bigint, "i64">("i64")
  static readonly i128 = new BorshSchema<bigint>("i128")

  // Floating point numbers:
  static readonly f32 = new BorshSchema<number, "f32">("f32")
  static readonly f64 = new BorshSchema<number, "f64">("f64")

  // Other scalars:
  static readonly bool = new BorshSchema<boolean>("bool")
  static readonly Unit = new BorshSchema<Unit>({ struct: {} })

  static Option<T extends BorshSchema<unknown, string>>(inner: T): BorshSchema<TypeOf<T> | null> {
    return new BorshSchema({ option: inner.schema })
  }

  static Array<T extends BorshSchema<unknown, string>>(
    inner: T,
    length: number,
  ): BorshSchema<TypeOf<T>[]> {
    return new BorshSchema({ array: { type: inner.schema, len: length } })
  }

  /**
   * Vec creates a vector (variable-length array) schema.
   *
   * Overloads:
   *  - For recognized numeric types (which we "brand"), the return type is a typed array.
   *  - Otherwise, it falls back to returning an array of the inner type.
   */
  static Vec(inner: BorshSchema<number, "u8">): BorshSchema<Uint8Array>
  static Vec(inner: BorshSchema<number, "u16">): BorshSchema<Uint16Array>
  static Vec(inner: BorshSchema<number, "u32">): BorshSchema<Uint32Array>
  static Vec(inner: BorshSchema<number, "i8">): BorshSchema<Int8Array>
  static Vec(inner: BorshSchema<number, "i16">): BorshSchema<Int16Array>
  static Vec(inner: BorshSchema<number, "i32">): BorshSchema<Int32Array>
  static Vec(inner: BorshSchema<number, "i64">): BorshSchema<Int64Array>
  static Vec(inner: BorshSchema<number, "f32">): BorshSchema<Float32Array>
  static Vec(inner: BorshSchema<number, "f64">): BorshSchema<Float64Array>
  static Vec<T>(inner: BorshSchema<T>): BorshSchema<T[]>
  static Vec(inner: BorshSchema<unknown, string> | BorshSchema<unknown>): BorshSchema<unknown> {
    let typedArrayType: string | undefined
    if (typeof inner.schema === "string") {
      switch (inner.schema) {
        case "u8":
        case "u16":
        case "u32":
        case "i8":
        case "i16":
        case "i32":
        case "i64":
        case "f32":
        case "f64":
          typedArrayType = inner.schema
          break
        default:
          break
      }
    }
    return new BorshSchema({ array: { type: inner.schema } }, typedArrayType)
  }

  static HashSet<T extends BorshSchema<unknown, string>>(inner: T): BorshSchema<Set<TypeOf<T>>> {
    return new BorshSchema({ set: inner.schema })
  }

  static HashMap<K extends BorshSchema<unknown, string>, V extends BorshSchema<unknown, string>>(
    key: K,
    value: V,
  ): BorshSchema<Map<TypeOf<K>, TypeOf<V>>> {
    return new BorshSchema({ map: { key: key.schema, value: value.schema } })
  }

  static Struct<T extends Record<string, BorshSchema<unknown, string>>>(
    fields: T,
  ): BorshSchema<{ [K in keyof T]: TypeOf<T[K]> }> {
    return new BorshSchema({
      struct: Object.fromEntries(
        Object.entries(fields).map(([key, schema]) => [key, schema.schema]),
      ),
    })
  }

  static Enum<T extends Record<string, BorshSchema<unknown>>>(
    variants: T,
  ): BorshSchema<EnumVariant<T>> {
    return new BorshSchema({
      enum: Object.entries(variants).map(([variantName, schema]) => ({
        struct: { [variantName]: schema.schema },
      })),
    })
  }

  /**
   * Serializes the given value into a Buffer.
   */
  serialize(value: T): Buffer {
    return Buffer.from(serialize(this.schema, value))
  }

  /**
   * Deserializes the given buffer.
   *
   * First, the third-party deserializer decodes the buffer.
   * If this schema was defined as a vector of a recognized numeric type
   * (as indicated by _typedArrayType) and the result is a plain array,
   * we convert that array into the corresponding typed array.
   */
  deserialize(buffer: Uint8Array): T {
    const result: unknown = thirdPartyDeserialize(this.schema, buffer)
    return this.convertToTypedArrays(result, this.schema) as T
  }

  private convertToTypedArrays(value: unknown, schema: Schema): unknown {
    // Handle null/undefined values
    if (value == null) {
      return value
    }

    // Handle array with _typedArrayType (Vec case)
    if (Array.isArray(value) && this._typedArrayType !== undefined) {
      switch (this._typedArrayType) {
        case "u8":
          return new Uint8Array(value)
        case "u16":
          return new Uint16Array(value)
        case "u32":
          return new Uint32Array(value)
        case "i8":
          return new Int8Array(value)
        case "i16":
          return new Int16Array(value)
        case "i32":
          return new Int32Array(value)
        case "i64":
          return new BigInt64Array(value)
        case "f32":
          return new Float32Array(value)
        case "f64":
          return new Float64Array(value)
        default:
          return value
      }
    }

    // Handle struct type
    if (
      typeof schema === "object" &&
      "struct" in schema &&
      typeof value === "object" &&
      value !== null
    ) {
      const result: Record<string, unknown> = {}
      const structFields = schema.struct

      for (const [key, fieldSchema] of Object.entries(structFields)) {
        const fieldValue = (value as Record<string, unknown>)[key]
        // Create a new BorshSchema instance for the field if it's an array type
        if (
          typeof fieldSchema === "object" &&
          "array" in fieldSchema &&
          !("len" in fieldSchema.array)
        ) {
          const innerType = fieldSchema.array.type
          if (typeof innerType === "string") {
            const tempSchema = new BorshSchema(fieldSchema, innerType)
            result[key] = tempSchema.convertToTypedArrays(fieldValue, fieldSchema)
          } else {
            result[key] = this.convertToTypedArrays(fieldValue, fieldSchema)
          }
        } else {
          result[key] = this.convertToTypedArrays(fieldValue, fieldSchema)
        }
      }
      return result
    }

    // Handle array type (fixed-length arrays or other array types)
    if (typeof schema === "object" && "array" in schema && Array.isArray(value)) {
      return value.map((item) => this.convertToTypedArrays(item, schema.array.type))
    }

    // Handle map type
    if (typeof schema === "object" && "map" in schema && value instanceof Map) {
      const newMap = new Map()
      const valueSchema = schema.map.value
      // Create a temporary schema for Vec value types
      let tempValueSchema: BorshSchema<unknown> | undefined
      if (
        typeof valueSchema === "object" &&
        "array" in valueSchema &&
        !("len" in valueSchema.array)
      ) {
        const innerType = valueSchema.array.type
        if (typeof innerType === "string") {
          tempValueSchema = new BorshSchema(valueSchema, innerType)
        }
      }

      for (const [k, v] of value.entries()) {
        const newKey = this.convertToTypedArrays(k, schema.map.key)
        const newValue = tempValueSchema
          ? tempValueSchema.convertToTypedArrays(v, valueSchema)
          : this.convertToTypedArrays(v, schema.map.value)
        newMap.set(newKey, newValue)
      }
      return newMap
    }

    // Handle enum type
    if (
      typeof schema === "object" &&
      "enum" in schema &&
      typeof value === "object" &&
      value !== null
    ) {
      const entries = Object.entries(value)[0]
      if (entries) {
        const [variantName, variantValue] = entries
        // Find the variant's schema
        const variantStruct = schema.enum.find((v) => variantName in v.struct)
        if (variantStruct) {
          // For Vec types, we need to create a new BorshSchema with the appropriate typedArrayType
          const variantSchema = variantStruct.struct[variantName]
          if (
            typeof variantSchema === "object" &&
            "array" in variantSchema &&
            !("len" in variantSchema.array)
          ) {
            const innerType = variantSchema.array.type
            if (typeof innerType === "string") {
              const tempSchema = new BorshSchema(variantSchema, innerType)
              return { [variantName]: tempSchema.convertToTypedArrays(variantValue, variantSchema) }
            }
          }
          // For other types, just process normally
          return { [variantName]: this.convertToTypedArrays(variantValue, variantSchema) }
        }
      }
    }

    // Return unchanged for primitive types or unhandled cases
    return value
  }
}

export type StructInput<T extends Record<string, BorshSchema<unknown>>> = {
  [K in keyof T]: TypeOf<T[K]>
}

/**
 * Helper type for enum variant discrimination.
 */
export type EnumVariant<T extends Record<string, BorshSchema<unknown>>> = {
  [K in keyof T]: {
    [KK in K]: TypeOf<T[K]>
  }
}[keyof T]

/**
 * Unit is defined as an empty record.
 */
export type Unit = Record<string, never>

export type StructFields = Record<string, BorshSchema<unknown>>
export type EnumVariants = Record<string, BorshSchema<unknown>>

// Re-export the TypeOf helper as "infer" if desired.
export type { TypeOf as infer }
