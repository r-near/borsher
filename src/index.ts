import type { Buffer } from 'buffer';
import { BorshSchema, type TypeOf } from './schema';

export function borshSerialize<T extends BorshSchema<unknown>>(
  schema: T,
  value: TypeOf<T>,
): Buffer {
  return schema.serialize(value);
}

export function borshDeserialize<T extends BorshSchema<unknown>>(
  schema: T,
  buffer: Uint8Array,
): TypeOf<T> {
  return schema.deserialize(buffer) as TypeOf<T>;
}

// Create a namespace (like Zod's 'z')
export namespace b {
  export type infer<T> = T extends BorshSchema<infer U> ? U : never;

  // Primitive types (lowercase with function calls)
  export const string = () => BorshSchema.String;
  export const u8 = () => BorshSchema.u8;
  export const u16 = () => BorshSchema.u16;
  export const u32 = () => BorshSchema.u32;
  export const u64 = () => BorshSchema.u64;
  export const u128 = () => BorshSchema.u128;
  export const i8 = () => BorshSchema.i8;
  export const i16 = () => BorshSchema.i16;
  export const i32 = () => BorshSchema.i32;
  export const i64 = () => BorshSchema.i64;
  export const i128 = () => BorshSchema.i128;
  export const f32 = () => BorshSchema.f32;
  export const f64 = () => BorshSchema.f64;
  export const bool = () => BorshSchema.bool;
  export const unit = () => BorshSchema.Unit;

  // Complex types (uppercase, direct references)
  export const Option = BorshSchema.Option;
  // biome-ignore lint/suspicious/noShadowRestrictedNames: <explanation>
  export const Array = BorshSchema.Array;
  export const Vec = BorshSchema.Vec;
  export const HashSet = BorshSchema.HashSet;
  export const HashMap = BorshSchema.HashMap;
  export const Struct = BorshSchema.Struct;
  export const Enum = BorshSchema.Enum;
}

export { BorshSchema, Unit } from './schema';
