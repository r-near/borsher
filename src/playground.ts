import { BorshSchema, type Unit } from './schema';

type Status =
  | {
      Pending: Unit;
    }
  | {
      Fulfilled: Unit;
    }
  | {
      Rejected: Unit;
    };

const statusSchema = BorshSchema.Enum({
  Pending: BorshSchema.Unit,
  Fulfilled: BorshSchema.Unit,
  Rejected: BorshSchema.Unit,
});

const status: Status = {
  Rejected: {},
};

const result = statusSchema.serialize(status);
console.log(result);
