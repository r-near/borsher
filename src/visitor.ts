import type { Schema } from "borsh"
import type { ArrayType } from "borsh/lib/types/types"

/**
 * Type guard for Vec schemas
 */
export function isVecSchema(schema: Schema): schema is { array: { type: string } } {
  return (
    typeof schema === "object" &&
    "array" in schema &&
    !("len" in schema.array) &&
    typeof schema.array.type === "string"
  )
}

/**
 * Get the element type of a Vec schema
 */
export function getVecElementType(schema: Schema): string | undefined {
  if (!isVecSchema(schema)) return undefined
  return schema.array.type
}

// Define valid TypedArray types
export type TypedArrayType = keyof typeof TYPE_CONVERTERS

type VisitorContext = {
  typedArrayType?: TypedArrayType
}

/**
 * TypedArray converters for Vec types
 */
const TYPE_CONVERTERS = {
  u8: (arr: number[]) => new Uint8Array(arr),
  u16: (arr: number[]) => new Uint16Array(arr),
  u32: (arr: number[]) => new Uint32Array(arr),
  i8: (arr: number[]) => new Int8Array(arr),
  i16: (arr: number[]) => new Int16Array(arr),
  i32: (arr: number[]) => new Int32Array(arr),
  i64: (arr: bigint[]) => new BigInt64Array(arr),
  f32: (arr: number[]) => new Float32Array(arr),
  f64: (arr: number[]) => new Float64Array(arr),
} as const

/**
 * SchemaVisitor implements the visitor pattern for traversing and transforming
 * Borsh schema structures. It handles the conversion of arrays to appropriate
 * TypedArrays based on the schema definition.
 */
export class SchemaVisitor {
  /**
   * Visit a schema node and transform its value according to the schema type
   */
  visit(value: unknown, schema: Schema, context?: VisitorContext): unknown {
    if (value == null) return value

    // Handle primitive types directly
    if (typeof schema === "string") {
      return value
    }

    // Visit different schema types
    if ("struct" in schema) {
      return this.visitStruct(value, schema)
    }
    if ("array" in schema) {
      return this.visitArray(value, schema, context)
    }
    if ("map" in schema) {
      return this.visitMap(value, schema)
    }
    if ("enum" in schema) {
      return this.visitEnum(value, schema)
    }

    return value
  }

  /**
   * Visit a struct schema and transform its fields
   */
  private visitStruct(value: unknown, schema: { struct: Record<string, Schema> }): unknown {
    if (typeof value !== "object" || value === null) return value

    return Object.fromEntries(
      Object.entries(schema.struct).map(([key, fieldSchema]) => [
        key,
        this.visit(
          (value as Record<string, unknown>)[key],
          fieldSchema,
          this.isVecSchema(fieldSchema)
            ? { typedArrayType: this.getVecElementType(fieldSchema) }
            : undefined,
        ),
      ]),
    )
  }

  /**
   * Visit an array schema and transform its elements
   */
  private visitArray(
    value: unknown,
    schema: { array: { type: Schema; len?: number } },
    context?: VisitorContext,
  ): unknown {
    if (!Array.isArray(value)) return value

    // Handle Vec types (arrays that should be converted to TypedArrays)
    if (!schema.array.len && context?.typedArrayType && context.typedArrayType in TYPE_CONVERTERS) {
      return TYPE_CONVERTERS[context.typedArrayType](value)
    }

    // Handle regular arrays
    return value.map((item) =>
      this.visit(
        item,
        schema.array.type,
        this.isVecSchema(schema.array.type)
          ? { typedArrayType: this.getVecElementType(schema.array.type) }
          : undefined,
      ),
    )
  }

  /**
   * Visit a map schema and transform its key-value pairs
   */
  private visitMap(value: unknown, schema: { map: { key: Schema; value: Schema } }): unknown {
    if (!(value instanceof Map)) return value

    return new Map(
      Array.from(value.entries()).map(([k, v]) => [
        this.visit(k, schema.map.key),
        this.visit(
          v,
          schema.map.value,
          this.isVecSchema(schema.map.value)
            ? { typedArrayType: this.getVecElementType(schema.map.value) }
            : undefined,
        ),
      ]),
    )
  }

  /**
   * Visit an enum schema and transform its variant value
   */
  private visitEnum(
    value: unknown,
    schema: { enum: Array<{ struct: Record<string, Schema> }> },
  ): unknown {
    if (typeof value !== "object" || value === null) return value

    const [variantName, variantValue] = Object.entries(value)[0] ?? []
    if (!variantName) return value

    const variantStruct = schema.enum.find((v) => variantName in v.struct)
    if (!variantStruct) return value

    const variantSchema = variantStruct.struct[variantName]
    return {
      [variantName]: this.visit(
        variantValue,
        variantSchema,
        this.isVecSchema(variantSchema)
          ? { typedArrayType: this.getVecElementType(variantSchema) }
          : undefined,
      ),
    }
  }

  /**
   * Check if a schema represents a Vec type
   */
  private isVecSchema(schema: Schema): boolean {
    return (
      typeof schema === "object" &&
      "array" in schema &&
      !("len" in schema.array) &&
      typeof schema.array.type === "string"
    )
  }

  /**
   * Get the element type of a Vec schema
   */
  private getVecElementType(schema: Schema): TypedArrayType | undefined {
    if (!this.isVecSchema(schema)) return undefined
    return (schema as ArrayType).array.type as TypedArrayType
  }
}
