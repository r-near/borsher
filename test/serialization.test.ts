import { describe, test, expect } from 'vitest';
import { b, BorshSchema } from '../src';

describe('primitive types', () => {
  test('string serialization', () => {
    const schema = b.string();
    const value = 'hello world';
    const buffer = schema.serialize(value);

    // Snapshot the binary data
    expect(buffer).toMatchSnapshot();

    // Also verify deserialization works
    const decoded = schema.deserialize(buffer);
    expect(decoded).toBe(value);
  });

  test('numbers', () => {
    interface TestCase<T> {
      schema: BorshSchema<T>;
      value: T;
    }

    const testCases: TestCase<number | bigint>[] = [
      { schema: b.u8(), value: 255 },
      { schema: b.u16(), value: 65535 },
      { schema: b.u32(), value: 4294967295 },
      { schema: b.u64(), value: 18446744073709551615n },
      { schema: b.i8(), value: -128 },
      { schema: b.i16(), value: -32768 },
      { schema: b.i32(), value: -2147483648 },
      { schema: b.i64(), value: -9223372036854775808n },
    ];

    for (const { schema, value } of testCases) {
      const buffer = schema.serialize(value);
      expect(buffer).toMatchSnapshot();
      expect(schema.deserialize(buffer)).toEqual(value);
    }
  });

  test('bool', () => {
    const schema = b.bool();

    const trueBuffer = schema.serialize(true);
    const falseBuffer = schema.serialize(false);

    expect(trueBuffer).toMatchSnapshot('true serialization');
    expect(falseBuffer).toMatchSnapshot('false serialization');

    expect(schema.deserialize(trueBuffer)).toBe(true);
    expect(schema.deserialize(falseBuffer)).toBe(false);
  });
});

describe('complex types', () => {
  test('Option', () => {
    const schema = b.Option(b.string());

    const someBuffer = schema.serialize('hello');
    const noneBuffer = schema.serialize(null);

    expect(someBuffer).toMatchSnapshot('Some value');
    expect(noneBuffer).toMatchSnapshot('None value');

    expect(schema.deserialize(someBuffer)).toBe('hello');
    expect(schema.deserialize(noneBuffer)).toBe(null);
  });

  test('Vec', () => {
    const schema = b.Vec(b.u8());
    const value = [1, 2, 3, 4, 5];

    const buffer = schema.serialize(value);
    expect(buffer).toMatchSnapshot();
    expect(schema.deserialize(buffer)).toEqual(value);
  });

  test('HashMap', () => {
    const schema = b.HashMap(b.string(), b.u32());
    const value = new Map([
      ['alice', 100],
      ['bob', 200],
    ]);

    const buffer = schema.serialize(value);
    expect(buffer).toMatchSnapshot();
    expect(schema.deserialize(buffer)).toEqual(value);
  });

  test('Struct', () => {
    const schema = b.Struct({
      name: b.string(),
      age: b.u8(),
      scores: b.Vec(b.u16()),
    });

    const value = {
      name: 'alice',
      age: 25,
      scores: [100, 95, 98],
    };

    const buffer = schema.serialize(value);
    expect(buffer).toMatchSnapshot();
    expect(schema.deserialize(buffer)).toEqual(value);
  });

  test('Enum', () => {
    const schema = b.Enum({
      Success: b.Struct({
        value: b.string(),
      }),
      Error: b.string(),
    });

    const successValue = { Success: { value: 'it worked!' } };
    const errorValue = { Error: 'something went wrong' };

    const successBuffer = schema.serialize(successValue);
    const errorBuffer = schema.serialize(errorValue);

    expect(successBuffer).toMatchSnapshot('Success variant');
    expect(errorBuffer).toMatchSnapshot('Error variant');

    expect(schema.deserialize(successBuffer)).toEqual(successValue);
    expect(schema.deserialize(errorBuffer)).toEqual(errorValue);
  });
});

describe('error cases', () => {
  test('invalid enum variant', () => {
    const schema = b.Enum({
      A: b.unit(),
      B: b.unit(),
    });

    // @ts-expect-error - Testing runtime error for invalid variant
    expect(() => schema.serialize({ C: null })).toThrow();
  });

  test('missing struct fields', () => {
    const schema = b.Struct({
      name: b.string(),
      age: b.u8(),
    });

    // @ts-expect-error - Testing runtime error for missing field
    expect(() => schema.serialize({ name: 'alice' })).toThrow();
  });
});

describe('complex nested structures', () => {
  test('game state example', () => {
    const schema = b.Struct({
      players: b.Vec(
        b.Struct({
          id: b.string(),
          score: b.u32(),
          inventory: b.HashMap(b.string(), b.u16()),
        }),
      ),
      status: b.Enum({
        Playing: b.unit(),
        Paused: b.unit(),
        GameOver: b.Struct({
          winner: b.string(),
          finalScore: b.u32(),
        }),
      }),
    });

    const value = {
      players: [
        {
          id: 'player1',
          score: 100,
          inventory: new Map([
            ['sword', 1],
            ['potion', 5],
          ]),
        },
      ],
      status: {
        Playing: {},
      },
    };

    const buffer = schema.serialize(value);
    expect(buffer).toMatchSnapshot();
    expect(schema.deserialize(buffer)).toEqual(value);
  });
});

describe('serialization compatibility', () => {
  test('classic BorshSchema syntax produces same bytes', () => {
    const newSchema = b.Struct({
      name: b.string(),
      age: b.u8(),
    });

    const classicSchema = BorshSchema.Struct({
      name: BorshSchema.String,
      age: BorshSchema.u8,
    });

    const value = {
      name: 'alice',
      age: 25,
    };

    const newBuffer = newSchema.serialize(value);
    const classicBuffer = classicSchema.serialize(value);

    expect(newBuffer).toEqual(classicBuffer);
  });
});

describe('async iteration', () => {
  test('async iteration', async () => {
    const arr = [1, 2, 3];
    // arr.forEach(async (i) => {
    //   await new Promise(resolve => setTimeout(resolve, 10));
    //   console.log(i);
    // });
    for (const i of arr) {
      await new Promise((resolve) => setTimeout(resolve, 10));
      console.log(i);
    }
  });
});
