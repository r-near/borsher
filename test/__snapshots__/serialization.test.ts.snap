// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`complex nested structures > game state example 1`] = `
{
  "data": [
    1,
    0,
    0,
    0,
    7,
    0,
    0,
    0,
    112,
    108,
    97,
    121,
    101,
    114,
    49,
    100,
    0,
    0,
    0,
    2,
    0,
    0,
    0,
    5,
    0,
    0,
    0,
    115,
    119,
    111,
    114,
    100,
    1,
    0,
    6,
    0,
    0,
    0,
    112,
    111,
    116,
    105,
    111,
    110,
    5,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > Enum > Error variant 1`] = `
{
  "data": [
    1,
    20,
    0,
    0,
    0,
    115,
    111,
    109,
    101,
    116,
    104,
    105,
    110,
    103,
    32,
    119,
    101,
    110,
    116,
    32,
    119,
    114,
    111,
    110,
    103,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > Enum > Success variant 1`] = `
{
  "data": [
    0,
    10,
    0,
    0,
    0,
    105,
    116,
    32,
    119,
    111,
    114,
    107,
    101,
    100,
    33,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > HashMap 1`] = `
{
  "data": [
    2,
    0,
    0,
    0,
    5,
    0,
    0,
    0,
    97,
    108,
    105,
    99,
    101,
    100,
    0,
    0,
    0,
    3,
    0,
    0,
    0,
    98,
    111,
    98,
    200,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > Option > None value 1`] = `
{
  "data": [
    0,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > Option > Some value 1`] = `
{
  "data": [
    1,
    5,
    0,
    0,
    0,
    104,
    101,
    108,
    108,
    111,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > Struct 1`] = `
{
  "data": [
    5,
    0,
    0,
    0,
    97,
    108,
    105,
    99,
    101,
    25,
    3,
    0,
    0,
    0,
    100,
    0,
    95,
    0,
    98,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`complex types > Vec 1`] = `
{
  "data": [
    5,
    0,
    0,
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > bool > false serialization 1`] = `
{
  "data": [
    0,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > bool > true serialization 1`] = `
{
  "data": [
    1,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 1`] = `
{
  "data": [
    255,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 2`] = `
{
  "data": [
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 3`] = `
{
  "data": [
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 4`] = `
{
  "data": [
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 5`] = `
{
  "data": [
    128,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 6`] = `
{
  "data": [
    0,
    128,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 7`] = `
{
  "data": [
    0,
    0,
    0,
    128,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > numbers 8`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    128,
  ],
  "type": "Buffer",
}
`;

exports[`primitive types > string serialization 1`] = `
{
  "data": [
    11,
    0,
    0,
    0,
    104,
    101,
    108,
    108,
    111,
    32,
    119,
    111,
    114,
    108,
    100,
  ],
  "type": "Buffer",
}
`;
